feat(docs): add keymap upgrader

joelspadin opened the pull request:
Added a documentation page with a script that upgrades deprecated key codes and behaviors to their replacements.

It does not attempt to make sure that it's only updating things in the bindings arrays, as properly parsing DT files is hard. It just uses regex word boundaries to make sure it's only replacing entire words.

Fixes #299

innovaker commented on docs/src/keymap-upgrade.js:
```diff
@@ -0,0 +1,55 @@
+      text = replaceWord(
+        text,
+        `&kp\\s+${deprecated}`,
+        `&none /* Cannot upgrade ${deprecated} */`
```
Perhaps something like:
`/* ${deprecated} no longer exists */`

joelspadin replied:
Done.

innovaker commented on docs/src/data/keymap-upgrade.js:
```diff
@@ -0,0 +1,75 @@
+  GUI: "K_CMENU",
+};
+
+export const Behaviors = {
+  "&cp": "&kp",
+};
```
Can we add this please?

`"&inc_dec_cp": "&inc_dec_kp"`

Thanks!

joelspadin replied:
Done.

innovaker commented on docs/src/keymap-upgrade.js:
```diff
@@ -0,0 +1,55 @@
+
+// The start of a behvioar like "&kp" isn't a word boundary,
+// so also match following the start of a line or whitespace.
+const WordBegin = "(?:\\b|^|(?<=\\s))";
+const WordEnd = "\\b";
+const UnimportantWhitespace = "( +)(?!\\S)";
```
Should we be using UpperCase?  Any thoughts @Nicell?

Nicell replied:
I personally don't use all upper case on `const`s in my JS code, but it definitely comes down to preference. I generally don't see many JS projects that use upper case for `const`.

innovaker replied:
Fair enough 👍  I'd not seen PascalCase used for this type of variable before which is why it caught my eye.

innovaker commented on docs/src/keymap-upgrade.js:
```diff
@@ -0,0 +1,55 @@
+  return text;
+}
+
+function replaceWord(text, word, replacement) {
+  // Match the word and all unimportant whitespace following it
+  const pattern = new RegExp(
+    `${WordBegin}(${word})${WordEnd}(${UnimportantWhitespace})?`,
+    "gm"
+  );
+  return text.replace(pattern, (match, word, spaceAfter) => {
+    return getAlignedReplacement(word, replacement, spaceAfter?.length ?? 0);
+  });
+}
```
I noticed it screwed up some of the ascii layouts that people use.  We could probably minimize that by adding rudimentary context wrapping to the RegExp.  Something along the lines of:
```
`(?:bindings\\s?=\\s?<[^>]+)${WordBegin}(${word})${WordEnd}(${UnimportantWhitespace})?(?:[^>]+>)`
```
It should handle `sensor-bindings` also.  What do you think please?

joelspadin replied:
I'd considered doing something like that, but if someone had this in their keymap, it'd break.
```
bindings = <
    /* < = > (whoops, the regex stops here)  */
    &kp LT &kp EQL &kp GT
>;
```
You'd need a smarter parser that's able to recognize comments, and while you can use regexes to build such a parser, regexes aren't suited for that on their own.

innovaker replied:
That's true, however:
- most (all?) shields have their ascii comments *above* the `bindings`, not inside them
- comments inside `bindings` containing `>` are likely to be rare
- it's better *not* to convert, than to convert unrelated parts of the file that the user doesn't anticipate or notice changing

Any thoughts?

innovaker replied:
The other approach to improve context is to include the behavior tags into the regexp, but it's probably more involved due to the numerous behaviors.

joelspadin-garmin replied:
I see three options here:
1. Do a simple word search and replace, which may convert unrelated parts of the file.
2. Use some naïve regex parsing to try to limit searches to specific parts of the file, which may fail to upgrade things that were supposed to be upgraded.
3. Use a proper parser.

In option 1, I believe the only things it might unexpectedly convert would be comments. Anything else would be seen by the preprocessor as an instance of one of the macros we're trying to upgrade, and thus should be upgraded too.

I personally see option 2 as worse because instead of unexpectedly converting some extra things that shouldn't have any effect on the function of the keymap, it might unexpectedly _not_ convert some things that do affect the function of the keymap.

Option 3 would obviously be the most reliable, but it is a lot harder. I tried that first using https://tree-sitter.github.io/tree-sitter/ but couldn't figure out how to resolve some ambiguities between the syntax for child nodes and properties. I could try porting Zephyr's DT parser to JavaScript instead of writing a parser from scratch.

joelspadin replied:
Whoops. I forgot I was still logged into my work account. If it wasn't already obvious, the above is also me.

innovaker replied:
Yeah, that was my impression from what you've said before.  But I tried it on many files and on balance I found that only stuff inside the bindings usually needed changing, and any comments that were also changed really threw me off.  I'd imagine the vast majority of users will be the same as most maps are rudimentary.

joelspadin replied:
For my own reference later, https://github.com/zephyrproject-rtos/zephyr/blob/master/scripts/dts/dtlib.py is Zephyr's parser. It is not trivial, but a port of that to JS would also be useful for making a VS Code (or similar IDE) extension for autocomplete in keymaps.

innovaker reviewed (changes requested):
Looks great @joelspadin!  Thanks for doing this 🥇 

Much of it is as I'd personally approach it.  Just some minor suggestions.

innovaker commented:
p.s. I didn't get a chance to actually *test* the builds of any converted keymaps, but hope to give some a go towards the end of the day and will compare before/after with git diff.

joelspadin commented on docs/src/keymap-upgrade.js:
```diff
@@ -0,0 +1,55 @@
+import { Codes, Behaviors } from "./data/keymap-upgrade";
+
+// The start of a behvioar like "&kp" isn't a word boundary,
```
Whoops. Fixed.

joelspadin commented on docs/docs/codes/keymap-upgrader.mdx:
```diff
@@ -0,0 +1,14 @@
+# Keymap Upgrader
+
+Many codes have been renamed to be more consistent with each other.
+Paste the contents of a .keymap file below to upgrade all deprecated codes to their replacements.
```
Done.

joelspadin commented:
Fixed the review comments and another issue I found where it wouldn't replace `<&cp` with `<&kp` because it didn't see `<&` as a word boundary.

innovaker commented:
Thanks for following-up @joelspadin.  My only concern is regarding modifying parts of the file that shouldn't be.  Please see my reasoning.

joelspadin commented:
I really hope this version works for other people, because I spent way too long fighting with tree-sitter to get it to load properly.

joelspadin commented on docs/src/keymap-upgrade.js:
```diff
@@ -0,0 +1,234 @@
+ * @returns {Parser.SyntaxNode | null}
+ */
+function findCapture(name, captures) {
+  for (const c of captures) {
```
Query API returns something that looks like

```ts
type Result = { 
  captures: { 
    name: string;
    node: Parser.SyntaxNode; 
  }[]; 
}[];
```
All the queries I'm running should just return results with a single capture, so this just finds and returns it.

joelspadin commented on docs/src/keymap-upgrade.js:
```diff
@@ -0,0 +1,234 @@
+ * Returns an empty array if no behavior was found.
+ * @param {Parser.SyntaxNode} paramNode
+ */
+function findBehaviorNodes(paramNode) {
```
The syntax tree for `<&kp A &kp B>` looks like

```C
(integer_cells // <
  (reference (identifier)) // (& (kp))
  (identifier) // (A)
  (reference (identifier)) // (& (kp))
  (identifier) // (B)
) // >
```

Given node "A" or "B", this walks back through siblings to find the previous `&kp` node, then walks forwards to find either the next `&kp` node or the end of the list. The list it returns includes everything from the previous behavior node up to but not including the next behavior node, which is the behavior and all its parameters.

joelspadin commented on docs/src/keymap-upgrade.js:
```diff
@@ -0,0 +1,234 @@
+  let currentIndex = 0;
+
+  for (const edit of edits) {
+    if (edit.startIndex < currentIndex) {
```
Given something silly like `&cp CURU`, we can both try to change `&cp` to `&kp` and `&cp CURU` to `&none /* "&cp CURU" no longer exists. */`. Those edits overlap so they can't both be applied.

That shouldn't happen in any reasonable keymap so I just made it throw out an arbitrary one of the edits instead of trying to do something smarter to resolve that.

joelspadin commented:
Added a bunch of comments with context since understanding what the new parser is doing isn't trivial.

Also see https://github.com/joelspadin/tree-sitter-devicetree/tree/main/test/corpus for some examples of how the parser sees things.

joelspadin commented:
Added a really hacky workaround for the problem where Firefox treats the path to tree-sitter.wasm as relative and Chrome et al. treat it as absolute. I just set up webpack to replace `"tree-sitter.wasm"` with `"/tree-sitter.wasm"` in tree-sitter.js.

joelspadin commented on docs/src/docusaurus-tree-sitter-plugin/index.js:
```diff
@@ -0,0 +1,39 @@
+        // which some browsers treat as absolute and others as relative.
+        // This breaks everything. Rewrite it to always use an absolute path.
+        rules.push({
+          test: /tree-sitter\.js$/,
```
I couldn't figure out how to get webpack to match exactly node_modules/web-tree-sitter/tree-sitter.js. Anything more specific I used just made it not do anything.

innovaker reviewed (changes requested):
Wow @joelspadin, you took this way further than I expected.  And shockingly quickly!

It LGTM, although I must admit I didn't completely understand the core system works.  I only had one suggestion.

Have you done much A/B testing?  i.e. diff-ing results as well as building them after?  If not, then I'll help you do some before we merge.

When it comes to merging, would you prefer the commits kept as they are, or squashed?

joelspadin commented on docs/src/keymap-upgrade.js:
```diff
@@ -0,0 +1,234 @@
+ */
+function applyEdits(text, edits) {
+  edits = sortEdits(edits);
+
+  console.log(edits);
+
```
Whoops. Yeah.

joelspadin commented on docs/docs/codes/keymap-upgrader.mdx:
```diff
@@ -0,0 +1,18 @@
+title: Keymap Upgrader
+sidebar_label: Keymap Upgrader
+hide_title: true
+---
```
Done.

joelspadin commented:
For some reason I can't reply directly to a couple comments, so

> When it comes to merging, would you prefer the commits kept as they are, or squashed?

Squashing is fine. I really only left it as two commits in case we found major issues with using WASM or something and I needed to revert to the regex method.

> I must admit I don't really understand this bit, but it seems to work and I trust your judgement.

If for example you put `CMMA` outside of the key bindings but still somewhere that the C preprocessor would see it as an identifier, it would have replaced it with the number for that key, which would probably make your keymap either fail to compile or at the very least not do what you intended. After the upgrade, it will be equally broken as before.

> Have you done much A/B testing? i.e. diff-ing results as well as building them after? If not, then I'll help you do some before we merge.

I have not done very extensive testing on this yet. So far I've just run a couple existing keymaps through it and then invented ones to test edge cases like that `&inc_dec_kp A CURU` gets turned into `&none`.

This also seems like it would be a good candidate for unit testing. Maybe we could pull in something like https://mochajs.org/ and add unit tests for the upgrader?

joelspadin commented:
After running this on a few more keymaps, should `CURU` be upgraded to `DOLLAR` now? Some of the existing keymaps used it and the comments above it showed it was used for `$`.

joelspadin commented:
Added minor fixes:
* Changed tab size to 4 in the input and output
* Updated parser to allow `&references` as node names (unlikely to affect keymaps)

innovaker commented:
> For some reason I can't reply directly to a couple comments, so

I've found that also.  Try doing it through the `Files changed` tab.

innovaker commented:
> After the upgrade, it will be equally broken as before.

Oh I think I understand now.  Well, you know the saying ... if it's ... broken, don't fix it 👍

innovaker commented:
> I have not done very extensive testing on this yet.

Will endeavor to do some later today.  It might also be worth calling in air support from our loyal testers.

innovaker commented:
> This also seems like it would be a good candidate for unit testing. Maybe we could pull in something like https://mochajs.org/ and add unit tests for the upgrader?

Can do!  Please don't feel obliged though given that the keymap upgrader will effectively be redundant in a few months (unless we make further code changes in the future).

innovaker commented:
> After running this on a few more keymaps, should CURU be upgraded to DOLLAR now?  Some of the existing keymaps used it and the comments above it showed it was used for $.

That's an excellent question.  I think we removed it simply because we weren't sure of its value.  But given that most of our initial users probably treated it as `DOLLAR` anyway, that's probably a good idea.  Might be best to use the `DLLR` abbreviation in this context?

innovaker commented:
@joelspadin, will pass it over to Pete for merging once you're happy and give me the green light.  Feel free to squash it yourself as it's one less thing for him to worry about.

joelspadin commented:
> Can do! Please don't feel obliged though given that the keymap upgrader will effectively be redundant in a few months (unless we make further code changes in the future).

Fair enough. I'll remap `CURU` to `DLLR` and then this should be ready (once it's been given enough testing). I've run about half the keymaps in this repo through it so far and the `CURU` thing is the only issue I've noticed so far, but I also don't have any of these keyboards to test with.

joelspadin commented on docs/src/data/keymap-upgrade.js:
```diff
@@ -0,0 +1,76 @@
+  PSTE: "K_PASTE",
+  VOLU: "K_VOL_UP",
+  VOLD: "K_VOL_DN",
+  CURU: null,
```
```suggestion
  CURU: DLLR,
```

joelspadin commented on docs/src/data/keymap-upgrade.js:
```diff
@@ -0,0 +1,76 @@
+  PSTE: "K_PASTE",
+  VOLU: "K_VOL_UP",
+  VOLD: "K_VOL_DN",
+  CURU: DLLR,
```
```suggestion
  CURU: "DLLR",
```

joelspadin commented:
Squashed and updated the page to indicate that it doesn't try to upgrade `#define`s.

innovaker commented:
@joelspadin, did we get many bites on the [testing](https://discordapp.com/channels/719497620560543766/719544230497878116/778667572945485834) (_sigh_, nobody posted on the issue)?  I lost track, what was the feedback please?

I've been busy myself, but will review #401.

joelspadin commented:
I think we only had one person who reported testing it and said it worked. Didn't see any other responses.

tominabox1 commented:
Tested on both of my boards and happy to report success on both. Cool feature!

joelspadin commented:
Rebased and added replacements for things like `MOD_RGUI` -> `RGUI` since there were apparently some keymaps still using that in `&mt`.

innovaker commented:
Tested thoroughly via #401.

@petejohanson, LGTM.

innovaker commented:
Awesome job @joelspadin! 👍   Thanks!
